sup/concurrency.cpp

```diff
@@ -25,7 +30,7 @@ namespace threading {
 //
 // Throws std::runtime_error:
 //  ARB_NUM_THREADS or OMP_NUM_THREADS is set with invalid value.
-util::optional<size_t> get_env_num_threads() {
+arb::util::optional<size_t> get_env_num_threads() {
```

Comment:
While we're here, this would be a good opportunity to reduce the use of `arb::util::optional` in public interfaces: this is going to be of a concern when we're in a mixed C++17 environment and we move to `std::optional`. (We're currently only using `optional` in one other public method, in `arb::mc_segment`).

Reply:
I could change it to return 0 or -1 on failure to read an environment variable. How does that sound.

I have no problem with making the fix, but I don't see what the problem is with keeping it.
When we move to C++17, wouldn't we just change it to `std::optional`?
If a user is currently using C++17, then how does our current set up cause a problem (besides potential confusion about the difference between `std::optional` and `util::optional`)?

Reply:
Isn't it possible to lift a name into a new namespace? Have an `arb::future`, where eventuall `arb::future::optional = std::optional`?

Reply:
The chief problem is one of interface: it's quite possible that we build Arbor with a C++17-capable compiler, but that users building against an installed version of it might want to use a different compiler with less robust C++ support (assuming all the ABIs are the same of course).

We should definitely move to `std::optional` when we can rely upon a C++17 build environment, but we shouldn't force a C++17 dependency through the interface, at least until compiler support is near universal.

Reply:
I think it is probably worth removing the `optional` from this interface, because it wouldn't be a massive loss (the error state is easy enough to encode in the return value with 0 or -1).
We might have a look at removing the `optional` on `mc_segment` so that this becomes a discussion we don't have to have.

If we were to keep `optional` in our interface (and similar types like `any`), I think Sam's idea of just using our `util::optional` until we move everything to C++17 is the way forward.
I want to avoid having multiple implementations of things in the library for different C++ standards: I would prefer to just implement everything in the minimum standard that we have selected for the library (currently C++14).